lib/recursive-methods.rb

```diff
@@ -1,49 +1,143 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n),
+#   where n is each 1 that can fit in the given number
+# Space complexity: O(n),
+#   because each n adds another method call to the stack
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError.new("#{n} is less than zero")
+  elsif n <= 0
+    return 1
+  else
+    return n * factorial(n-1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n/2),
+#   where n is the number of characters in the string,
+#   but /2 because you're chopping off two characters each time
+# Space complexity: O(n/2)
```

Comment:
1.  We drop the coeficient.
1.  Since you're creating a new array with `s[1...-1]`, the time/space complexity is actually O(n<sup>2</sup>)